### PR TITLE
[DRAFT] Feature: Dynamically Shaded Sprites

### PR DESCRIPTION
**Pull request by Koranir**
## Feature Details
This PR adds a system for dynamically shadowing sprites based on a normal map and (optional) base colour and emission maps. It gets lit by stellarObjects with the lightcolor attribute.

## UI Screenshots

https://user-images.githubusercontent.com/101683475/202168879-c388c514-5597-4d6b-9278-50c823370f52.mp4

image

## Usage Examples

Ideally, the ship's data would have all 3 maps in addition to the original sprite.
```
ship "Emerald Sword"
	name "Big Boss"
	sprite ship/emeraldsword/emeraldsword
	normal ship/emeraldsword/emeraldswordnormal
        base ship/emeraldsword/emeraldswordbase
	emit ship/emeraldsword/emeraldswordemit
	thumbnail thumbnail/emeraldsword
	uuid 5680387a-ad51-4599-a052-63f03d873564
	attributes
		category "Heavy Warship"
		cost 100000000
		mass 1520
```

However, the base colour and emmission can be ommitted, and an approximation of what they would be will be used. **This is only viable if the ship is symmetrical, or close to**. Ships like the bactrian or asteroids need individual base color maps.
```
ship "Emerald Sword"
	name "Big Boss"
	sprite ship/emeraldsword/emeraldsword
	normal ship/emeraldsword/emeraldswordnormal
	thumbnail thumbnail/emeraldsword
	uuid 5680387a-ad51-4599-a052-63f03d873564
	attributes
		category "Heavy Warship"
```

StellarObjects can have a light attached to them. 
The colour and intensity of the lights are controlled by a lightcolor attribute
The light intensity diminishes over distance.
```
	object "Ember Reaches"
		sprite planet/wormhole-red
		distance 2354.54
		period 1371.69
		lightcolor 1. .2 .1 200.
```

The light the system's star gives off can also be changed
```
system Coluber
	starcolor 1. .87 .79 1.1
```

There are a few example images in the files changed of what the maps look like.

The normal map is neccessary for the advanced shader to work, but the shader's behaviour will revert to normal if none is supplied.

### TODO:
- [x] Support multiple lights with variable intesity and positions, not just the center of the system (Partially done)
- [ ] Add light radius to reduce errors
- [ ] Make quality settings to reduce bright pixels
- [ ] Tune the contrast and brightness
- [x] Compatability with motion blur
- [x] Make asteroids and flotsam use this system (<-- GONNA BE GOOD)
- [ ] Update comments (right now they're all copied from SpriteShader.cpp)
- [ ] Add lights to every wormhole
- [ ] Update ships to all use normal maps
- [ ] Update asteroids to all use normals
- [ ] Update XCode project

### To make normal maps:

1. Download [This .blend](https://drive.google.com/file/d/1Je91RR8BVd1KmxtisAk2NeVA7kCARvOG/view?usp=share_link).
2. Open the .blend of the body you want the normal of.
3. Append the 'Normal' object from the downloaded blend into the .blend of the body.
4. Select the body and then select the appended object.
5. Use ctrl+l and select link material
6. If the camera is rotated, go to the shaders tab and edit the 'angle' input value of the normal material.
7. From now, rendering the image or animation shold output a normal map.

## Performance Impact
I haven't tested too much, but most of the work is pretty light so there shouldn't be a big performance hit.

### CODE REVIEW

**Comment by Zitchas** [comment]
That looks awesome!

That also sounds like a process that might be at least partially automatable.

**Comment by quyykk** [comment]
See also https://github.com/endless-sky/endless-sky/pull/3033

**Comment by Saugia** [comment]
^ Seconding quyykk's referral there. This seems to have been attempted and heavily discussed in the past but was ultimately shot down by the game's creator, MZ. There are a lot of pros and cons to it too, but I'd have to agree with the cons and concerns listed in that PR. Regardless, we'll see what others have to say.

**Comment by Amazinite** [comment]
Just a heads up that whether or not this feature gets merged, irrespective of how you implement it, is on a knife's edge. As was noted the other week, this isn't the first time this has been attempted, and the concerns back then weren't implementation related. The concerns are moreso a matter of this increasing the work required to create new artwork for the game, a concern which exists no matter which way shading like this is done.

**Comment by Saugia** [comment]
My opinion on the matter.

First of all, thank you for your hard work you've put into this. I know it's been a lot so far. Dymanically shading sprites is always an interesting method that 2d space games can have as possibilities.

However, I don't think Endless Sky is that game.

The shading looks extremely dull. Our ships are not designed with dynamic shading in mind. The way it looks is way too flat and bland. We do quite a lot of accenting, adjusting and exaggerating of shadows, textures, etc. in post-production in Gimp and Blender. That exaggeration is what brings are ships to life. While dynamic shading is more realistic, Endless Sky is not a game about realism. I feel this sort of look is better for a different game.

Implementation would be a serious ton of work. We are talking about hundreds of ships. Not only that, but we would have to re-do every post production entirely (scratches, rust, and other textures). Hardpoints are also a concern, as they here are not shown to follow the shading. I wouldn't want one doing this without the other.

My largest concern is ease of access to new contributors. This is quite a lot of intricacy for someone brand new to Endless Sky artwork, and comes of as extremely daunting. It also narrows our creators' platform usage. This strongly prefers Blender 3.0+ Cycles, where we still have artists (including one of our lead artists who has done half of the game basically) working in earlier versions of blender, and not comfortable moving up. They, myself, and others prefer the traditional gimp post-processing over blender texture nodes as we feel it is the traditional look of Endless Sky. I feel with dynamic shading, we lose this.

In conclusion, this would require a huge commitment, and based on what I said above, I don't think this is appropriate for Endless Sky.

**Comment by Zitchas** [comment]
I'm neither a code nor graphics person, so I can't comment on how hard this would be to implement or support. That being said, I very much like the looks of it, and would definitely appreciate having it in game. The EV games had shadows, albeit using sprite sheets, not shaders, and that added a depth to the game that I think ES lacks. So I'd love to have it.

But if the art and code people say it's too hard... Well, they're the ones that have to take this and make it work for years to come.

**Comment by beccabunny** [comment]
Gonna be honest, i wanted to see something like that for a lot of time in ES; but if it means redoing most of the post processing for every ship, I'm not sure if it can really be the way to go. I for sure wouldn't want to remake again the 50 or so ships i have in game or ready to be added.
One thing i would want to have preserved if it can work without redoing sprites is the red lightning near wormholes and other objects, as it doesn't seem fully dependent on the dynamic shading and looks damn cool

**Comment by Zitchas** [comment]
Perhaps worth noting is that there's been a trend over the past year of moving away from post-processing towards proper materials. I've been pushing for it for years so that we can drop ships into scenes cleanly without having to fiddle for hours trying to manually recreate a 2D post-processing layer on a 3D scene of the ship engaged in action of some kind. 

So I kind of view having our ships upgraded to that as (hopefully) inevitable. It might not be this year, it might not be next year, but I suspect that it is only a matter of time until someone does it. Just like Becca upgraded all the old ships to the beautiful new ones, I think it likely that someone will do the same to bring everything entirely into blender. All it takes is one person who really grasps materials.

As far as that goes, I still think that the process of making the normal map looks like something that should be largely automatable. If someone can figure that out, a significant amount of the workload for this will be taken care. Has anyone here done any automation like this before?

**Comment by 1010todd** [comment]
I would really like to see this implemented though maybe exclude ships from this PR for now if that's controversial? At least it could be PRed and reviewed later. Sure it's more work but it's an improvement over the static light-per-ship we got which can look jarring sometimes, we can't get anywhere without doing something. If anything, with this feature available first it would be easier for people to slowly revisit ships and improve it with the dynamic shading in mind rather than doing everything in big chunk which creates a bunch of problem for us.

Honestly I don't feel like we should take MZ's method of overlaying grunges and manually painting AO as "style", he just didn't have time to learn Blender and only did what's quicker for him, iirc he also said himself that he didn't texture in Blender because he thought it wouldn't worth the effort (which is totally fair, since he was doing the codes and storyline as well). Anyway, to me it feels like it was initially done more as fast and convenient method that looks sufficient than deliberate art style choice for the look. Hardpoint might be a concern but it's already an existing problem so I don't see this as enough of a problem to block this PR. From the method described it doesn't appear like you *need* Blender 3.0+ for normal map generation either, you only need Cycles engine for a bit of node stuffs and that exist in Blender 2.79. But one thing I have to agree with Saugia that this will raise the barrier of entry quite a lot and it's not entirely just the process of making the normal maps but just their existence in the folders would look way more intimidating than just some ship PNGs. I don't really see plausible way to really fix this but since it appears like some of the code is literally written as string/text perhaps it could be made to read text file from plugins so at least dynamic lighting can be enabled with plugin and be an option for those who want it.


Anyway, about the code of this PR I recall that the game will try to generate collision mask for every image in ship and asteroid folder and I don't see anything in this PR that touch that yet. We might want to skip collision mask generation for normal maps and such.

**Comment by Zitchas** [comment]
Question for the code people: How hard will this be to maintain? How much overhead? Is all of this something that's going to make everything a bit harder all the time? Or is it really only going to affect anything when you deal with something specifically related to how ships are drawn on the screen?

What I'm getting at is that we have plenty of features that we support that are, at least initially, "plug-in only." If these dynamic shaders just make the display-ships section more complicated, but doesn't add any burden to the rest of the code (AI and how ships work seems to be the main areas that come up); then this seems like something that would be good to integrate. Plug-ins would then have the option of having dynamic shaders; and those interested in doing so could slowly build up an official "dynamic shading plug-in" alongside high-dpi that is gradually filled up over time. 

Sure, it might take a couple years. But then, it's taken a couple years for the current ship upgrades to take place, and I suspect starting with the current updated blender models and going the step farther with materials is probably quite a bit faster than rebuilding everything from scratch as had to be done with a lot of the original models. From what I gather, once one has figured out a good set of materials for one ship in a shipyard, it isn't a huge stretch to use the same materials for other ships in the same "set."

**Comment by RisingLeaf** [comment]
> Question for the code people: How hard will this be to maintain? How much overhead? Is all of this something that's going to make everything a bit harder all the time? Or is it really only going to affect anything when you deal with something specifically related to how ships are drawn on the screen?
> 

Because our shader system is old and build kind off inefficient it wouldn't necessary be more work but more complexity. The concerns here are more from the art people.

**Comment by Saugia** [comment]
After some consideration, I want to address my and other's concerns before completely writing off this PR. For what it's worth, I'm entirely for this implementation if it doesn't mean we have to redo everything for it, especially right away.

First and foremost - your choice of an example system makes it kind of hard for me to seethe shading, with all of the effects and asteroids flying everywhere. Could you please have several footage examples of flying around various types of stars, preferably in quieter environments? Fly in a way that circles the star and shows all shading directions, and then close and far away to show those too. A large part of my concern visually is coming from this because it doesn't look like the shading on the ship itself is very good at showing depth which we exaggerate a bit using post. Please use an Osprey again.

I will agree with Todd that the lighting style is lazy, but the post style of textures feels better for me for theses sorts of models. It's also much simpler to do versus texture nodes, and the visual difference in the game isn't that noticeable for non-artists. I have to disagree with Zitchas regarding scenes - it's just as easy to use post texturing to do them, actually easier. While texturing in Blender is a good overall practice, it's not the "100% way to go", "inevitability". I would only agree with this if the process to create ships for this shading becomes just as easy as the way we do it now. We also will likely never have *footage* for scenes as that's several steps beyond this and heavily relies on performance.

Secondly, my larger concern is as everyone is mentioning, the bar to entry. This normal map system will be quite a large load to new artists as opposed to simply layering textures. If we *can* get an automated process, and it is *without flaw*, perhaps that could work, but having an extensively written out guide would also be desired. I don't think this is worth committing to if it means new creators are turned away by the task.

I agree with Becca that the conditional lighting based on objects though, it's actually something I really like. If we decide to wait on this again, we could keep that. 

Finally, if my concerns are alleviated, I would have to put a stance that this sort of thing waits along with the Map Rework for a more complete base game. I don't like the idea of rushing it into things while we're half-way into them. I know that means future ships and such can then be done for the new dynamics, but re-doing every single ship will take a lot of time. I would prefer that happen all at once (aka in another PR in a few years or such), rather than get everyone mixed up in it now and taking attention away from progressing in beta.

**Comment by Amazinite** [comment]
Given that this PR is still in draft form, and because there are still major outstanding question on whether it will even be merged, I'm closing this PR. And admittedly, for the many reasons mentioned, I think that this feature is one that we'd be hard pressed to ever merged. Not only would it be a big change in how new artwork needs to be made for ships, but also a change in the game's graphical style for said ships. I'm completely open to other forms of new shader support (e.g. #8400, or that black hole video I saw), but this feature might not be one of them.
